flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java

```diff
@@ -376,7 +372,7 @@ private void handleStartTaskExecutorServicesException(Exception e) throws Except
 
 		return FutureUtils
 			.runAfterwards(
-				taskCompletionTracker.failIncompleteTasksAndGetTerminationFuture(),
+				taskSlotTable.freeAllSlots(new FlinkException("The task executor is shutting down.")),
```

Comment:
The current way of ensuring that all tasks are being terminated before the returned future is completed is super hard to understand because the code path is too convoluted and too many components depend on each other to make it happen. It took me actually quite some time to figure this out and I'm somewhat familiar with the code. Instead of introducing additional APIs which need to be called in a very specific manner, I would suggest that we simply call `taskSlotTable.closeAsync()` which returns a future which is completed once all slots are closed. A slot is closed if all its task have terminated. The information about a task being closed does not need to be passed into the `TaskSlotTable` via the `freeSlot` but we could simple take a look at the `Task#getTerminationFuture` in order to control when a `TaskSlot` is completely released.

A good design principle is context independence. If possible a component should work independent of its context meaning that it tries to decouple it self as good as possible from other components. In this case, for example, we require that `TaskExecutor` calls `TaskSlotTable#freeSlot` after we have called `TaskSlotTable#freeAllSlots` and before we can call `close`.

Reply:
Indeed, it is complicated. I tried to avoid changing too much the process of freeing slots. At the moment it involves quite some steps:

- JM calls freeSlot and puts the slot into the releasing state
- task update state final (TE)
- remove task (table)
- if slot is in releasing state and it has no running tasks
  - the async table.freeSlot and
  - notifies RM/close JM connection (slot actions impl in TE)

It would be nice that `table.freeSlot` would not have to go through TE but I guess somehow `table.freeSlot and notifies RM/close JM connection` has to be scheduled async into the main thread atomically.

True, it would be nicer if task management was part of the slot as task belongs to it then it could be tested only there and slot freeing/closing would guarantee interruption, now it is mixed into the table class.

The first approach in this PR also kept slot/table not dealing with threading and futures leaving TM taking care of it. I agree that the suggested approach separates better responsibilities w/o changing the existing process of freeing slot too much and introducing async behaviour to slot/table should not be a big problem now.

I implemented the suggested approach with `closeAsync`. Some final cleanup of slot/table (timeService.stop and  memoryManager.shutdown) happens in sync callbacks to the termination of tasks w/o scheduling it to the main thread of TM. Also, the final cleanup of table internal data structures after slot freeing potentially happens after the closing future but better to be scheduled before. Although this does not look implicitly a problem now, it is better to rethink more the design and responsibilities by the next opportunity and keep all actions thread-confined to avoid accidental putting of critical stuff into the callbacks w/o proper synchronisation. As discussed, one solution could be that slot/table get main thread executor in future and become more independent from TM.